src-java/rule-manager/rule-manager-implementation/src/main/java/org/openkilda/rulemanager/adapter/PersistenceDataAdapter.java

```diff
@@ -0,0 +1,125 @@
+import org.openkilda.model.PathId;
+import org.openkilda.model.Switch;
+import org.openkilda.model.SwitchId;
+import org.openkilda.model.SwitchProperties;
+import org.openkilda.model.TransitVlan;
+import org.openkilda.persistence.PersistenceManager;
+import org.openkilda.persistence.repositories.FlowPathRepository;
+import org.openkilda.persistence.repositories.RepositoryFactory;
+import org.openkilda.persistence.repositories.SwitchPropertiesRepository;
+import org.openkilda.persistence.repositories.SwitchRepository;
+import org.openkilda.persistence.repositories.TransitVlanRepository;
+import org.openkilda.persistence.repositories.VxlanRepository;
+import org.openkilda.rulemanager.DataAdapter;
+
+import lombok.Builder;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+public class PersistenceDataAdapter implements DataAdapter {
+
+    private final FlowPathRepository flowPathRepository;
+    private final SwitchRepository switchRepository;
+    private final SwitchPropertiesRepository switchPropertiesRepository;
+    private final TransitVlanRepository transitVlanRepository;
+    private final VxlanRepository vxlanRepository;
+
+    private final Set<PathId> pathIds;
+    private final Set<SwitchId> switchIds;
+
+    private Map<PathId, Flow> flowCache;
+    private Map<PathId, FlowPath> flowPathCache;
+    private Map<PathId, FlowTransitEncapsulation> encapsulationCache;
+    private Map<SwitchId, Switch> switchCache;
+    private Map<SwitchId, SwitchProperties> switchPropertiesCache;
+
+    @Builder
+    public PersistenceDataAdapter(PersistenceManager persistenceManager,
+                                  Set<PathId> pathIds, Set<SwitchId> switchIds) {
+        RepositoryFactory repositoryFactory = persistenceManager.getRepositoryFactory();
+        flowPathRepository = repositoryFactory.createFlowPathRepository();
+        switchRepository = repositoryFactory.createSwitchRepository();
+        switchPropertiesRepository = repositoryFactory.createSwitchPropertiesRepository();
+        transitVlanRepository = repositoryFactory.createTransitVlanRepository();
+        vxlanRepository = repositoryFactory.createVxlanRepository();
+
+        this.pathIds = pathIds;
+        this.switchIds = switchIds;
+
+        encapsulationCache = new HashMap<>();
+    }
+
+    @Override
+    public Map<PathId, FlowPath> getFlowPaths() {
+        if (flowPathCache == null) {
+            flowPathCache = flowPathRepository.findByIds(pathIds);
+        }
+        return flowPathCache;
+    }
+
+    @Override
+    public Flow getFlow(PathId pathId) {
+        if (flowCache == null) {
+            flowCache = flowPathRepository.findFlowsByPathIds(pathIds);
+        }
+        return flowCache.get(pathId);
+    }
+
+    @Override
+    public FlowTransitEncapsulation getTransitEncapsulation(PathId pathId) {
+        if (encapsulationCache.get(pathId) == null) {
+            encapsulationCache = new HashMap<>();
```

Comment:
```
getTransitEncapSulation("id1") 
getTransitEncapSulation("id2")
```

Second call will remove information about "id1" from cache because `encapsulationCache.get("id2")` is null and cache will be set to empty set